### ISSUE TITLE
AIs do not switch target when hit by a character they aren't aggro'd to

### ISSUE THREAD

**Issue by eotetb** [aura-project/aura#456]
I have seen this happen a few times now, when my party member hits a mob that has already aggro'd me, instead of the mob switching their target to the other party member, the continue going after me.

One example of this is if a party member hits a mob with defense loaded, the monster than retaliates and hits the user that originally aggro'd them instead of switching to the new target. A second example is if a party member starts a combo but finishes midway through, the monster will immediately retaliate against the user it originally aggro'd instead.

**Comment by Sairii** [comment]
Actually I do not know if we even have aggro switching stuff yet.
I think thats part of the new fancy AI server thingy that exec has plannedish?

**Comment by exectails** [comment]
Actually, I thought we had that, because it actually requires a check to prevent that from happening. Looks like we did add that check at come point though, possibly unintentionally.

```csharp
public override void Aggro(Creature target)
{
	if (this.AI == null)
		return;

	// Aggro attacker if there is no current target,
	// or if there is a target but it's not a player, and the attacker is one,
	// or if the current target is not aggroed yet.
	if (this.Target == null || (this.Target != null && target != null && !this.Target.IsPlayer && target.IsPlayer) || this.AI.State != AiScript.AiState.Aggro)
		this.AI.AggroCreature(target);
}
```

Did we ever research this, @Sairii? I remember situations where monsters would not stop attacking one player, while in other situations they would. Isn't it normal for them not to switch? Or maybe it's based on who did the most damage/hits?

**Comment by Sairii** [comment]
There are these Aggro scales in some of the mobs. Most of the time boss monsters.
It is based on damage with a multiplier / type of skill / attack.

**Comment by exectails** [comment]
Yea, I remember those of course. Question is what the default is I guess.

**Comment by Sairii** [comment]
That aggro list makes sense imho.

melee does most aggro (depending on dmg)
magic and range are 2nd place (depending on dmg)

There might be a default somewhere or maybe that is the default.
Anyways it should be easy enough to test this.

**Comment by eotetb** [comment]
I feel like I should do some testing on official myself to see how aggro should act. It seemed to me that triggering defense at the very least did cause mob aggro to switch targets. Something just felt a little off to me in Aura. Its a little hard to remember when we tend to one shot everything with crazy AoE skills on official now. 

Also in regards to the aggro scales @Sairii posted, does "supporthit" refer to support shot specifically?

**Comment by Sairii** [comment]
Yes, that was my guess. Support shot generates less aggro..... was what i thought....

but i already tested on official and it is totally random.....
- some mobs have delays for aggro switching
- apparently even bosses switch ALWAYS even if there has been only little damage from the other person
- and also it kinda does not always switch?!??!?!?!?

Anyways... I tested that stuff with my pet, which might not really work the way it works with other players.
I tested it on ancients (caus they wouldnt die from 1 hit) and later ciar beginner on golem.

Then I gave up after I saw golem just herpaderping.

**Comment by exectails** [comment]
Wasn't Support Shot supposed to not generate *any* aggro? Wasn't that the whole point of it? Although, I remember hearing that the skill was bugged, and not working as intended. Maybe players just misinterpreted a lowered aggro weight instead of none as being a bug?

>why did they have an extra option for dmg...

Do you mean the last one, that only says `damage`? I would expect that to be normal damage (melee), while the others are specific sub-types, coming from magic skills, ranked skills, clean hits, or Support Shot (?).

It seems a little weird that Support Shot would have the same weight as normal damage at first, but I guess it actually makes sense, because Ranged, which would have four times the weight of a melee attack, would draw the enemy away basically every time.

**Comment by boilerplate5** [comment]
It's definitely a lowered aggro. Support Shot's main draw is that it doesn't affect the knockback gauge at all.

**Comment by Sairii** [comment]
What is clean hit then?
I am pretty sure clean hit is supposed to be melee aggro
http://wiki.mabinogiworld.com/view/Types_of_Hits

and damage is just damage that scales on top?

**Comment by exectails** [comment]
A clean hit is when you hit the enemy and "Clean Hit" appears on the screen. I don't know when exactly that is, and I wouldn't trust the wiki to tell me =<

>and damage is just damage that scales on top?

What do you mean? If you hit an enemy for 20 damage with 0.2 weight (Ranged Attack), it's +4 for the "aggro scale". If you hit them for 20 damage with 0.05 weight (normal attack, Combat Mastery), it's +1 for the scale.

**Comment by Sairii** [comment]
i think only melee hits are clean hits?

also
>1000 (dmg) x 0,05 (dmg) x 0,2 = 10 // magic
>400(dmg) x 0,05(dmg) x 0,5 = 10 // melee ... would make sense?
is what i was thinking.

And yes it kinda doesnt make sense that way because only ranged / melee / magic would be sufficient

**Comment by eotetb** [comment]
I just ran Coill to see how aggro changes when specific skills are triggered.

For defense, every monster I tried this with (either mob aggro'd pet and I triggered defense or vice versa) would switch aggro immediately. 

On the other hand activating an enemies counterattack did not cause aggro to change (mob aggro'd me, I triggered its counter, mob proceeded to keep aggro on pet and walk around it with defense loaded).

Also, since defense seemed to always cause aggro to switch I was surprised when triggering heavy stander did not always cause enemy aggro to switch. A Gorgon had aggro'd my pet, I used a normal attack on it and triggered its heavy stander. It waited about a second, but it then proceeded to continue attacking my pet instead of me.

I was definitely wrong on mobs switching aggro when stopping mid combo on them. I had mobs aggro'd to my pet, I hit them once and they continued attacking my pet. The mob did pause first before continuing its attack on my pet though. In aura monsters seem to retaliate with no delay if a party member hits a mob and stops mid combo. That might have just been from their AI though. Snake in official vs a metal skeleton in aura.

The last thing I noticed though was that the Giant Headless in the boss room also wasn't switching its aggro state immediately after getting attacked.  There were times where it was aggro'd on to me, my pet would do a full regular attack combo and then it would continue to attack me rather than switch, much like a normal monster.  

I didn't pay too much attention to the damage I was doing, but I was doing around 500-700 without critical hits and around 20 for my pet.

**Comment by Sairii** [comment]
>Are we sure the Kobold actually switched? He ran into your general direction, so he might've been circling you, but it might as well have been fleeing or wandering.

Yes I only look for the '!!' because of the delay and OnHit events.

>For defense, every monster I tried this with (either mob aggro'd pet and I triggered defense or vice versa) would switch aggro immediately. 

Did I miss something? Since when does defense produce aggro?
I have to test that myself. Maybe it is because of aggro on battlestance.
Also we do not know yet if every mob actually uses the aggro chart that I posted.

**Comment by exectails** [comment]
Oh, you're right. Missed the exclamation marks behind the chat bubble.

>Did I miss something? Since when does defense produce aggro?

If you hit a monster that has Defense active?

**Comment by eotetb** [comment]
Yes, I'm referring to if the monster has defense loaded and you hit them while its loaded.

**Comment by Sairii** [comment]
OHHHH
ye but thats no "aggro" thats an OnDefense event.
You dont get the '!!' for that.

**Comment by exectails** [comment]
So they hit the creature that hit their Defense and then they continue battling their actual target?

**Comment by Sairii** [comment]
Yes, I just tested that.
They hit the target that triggered the defense and then continue with their original target.

**Comment by eotetb** [comment]
If that's the case for defense, then I suppose there is currently an issue with defense in that regard in aura. That behavior does not happen in aura, mobs will retaliate against the user they originally aggro'd instead of the user that him them while defense was active. 

Should I create a new issue for defense, since this seems to be different from the aggro enhancement?

**Comment by exectails** [comment]
No, this is all basically one aggro problem. Aura has a very simple "that's my target, that I'm gonna attack" system, switching in any way, even temporarily, wasn't intended originally =/

**Comment by Sairii** [comment]
We still need tests regarding this :<

**Comment by exectails** [comment]
We do... and we need a way to handle this temp switching.
